### ISSUE TITLE
Has anybody successfully implemented constriction with the viscous solver?

### ISSUE THREAD

**Issue by btbest** [DamCB/tyssue#246]
I've just about tried everything I can think of, but at this point I'm about to discard the viscous solver and settle that I just can't have a nice model that runs based on forces and time rather than from equilibrium to equilibrium.

But before that, I wanted to ask if anyone actually has managed to implement constriction with this. I've been trying to make FaceContractility work, but even with my attempts at rewriting its gradient function, it just doesn't seem to want to produce anything that resembles constriction of a face, or anything that resembles what its energy function produces with the quasistatic solver. I'm specifically trying to get constriction (and invagination) of a "mesoderm" on the ellipsoid model that tyssue was first published with, but I'd be interested in anything functional where any surface of anything constricts.

Or even more generally, is anyone using the viscous solver for anything at all? What sort of behaviours have you managed to implement?

**Comment by sniffo** [comment]
Hey, trying the email response on here for the first time so apologies for
any issues. I have been able to implement the quasistatic and viscous
solvers to simulate various forces and such. What exactly is the issue you
are facing? Most of the time any issues with the Euler Solver is related to
the timestep and the actual physical parameters used. I have mostly done
simulations in 2D though, so the ellipsoid geometry might have additional
parameters like lumen elasticity to consider. Do you have any notebooks
etc. that can show what you have thus far?

Best,
Tasnif

**Comment by btbest** [comment]
Can you give me some sample parameters that you use? As for step sizes, I've tried solving with step sizes from 0.1 to 1 and have stuck with 1 because it allows me to follow the calculations more easily.

In the bigger ellipsoid setting, no matter what parameters or combination of effectors I've tried (including FaceContractility alone, no other effectors such as volume elasticity), the system always ends up spiralling to infinity with vertices oscillating at higher and higher amplitude perpendicular to the face's surface plane.

I've taken a microsystem with three hexagons out of one of the tutorial notebooks to look at how the 2D effectors work, for example like shown below. It clearly doesn't produce isotropic constriction, though I don't know if this has anything to do with how it acts up in the ellipsoid setting. Of course it could be that it's not even supposed to be isotropic. It just always seemed to me that the energy function does produce isotropic constriction, and hence I get the feeling that the gradient function of this effector doesn't match its energy function.

I got here in the process of inspecting each of the ellipsoid model's effectors. Since the oscillations were perpendicular to the face's surfaces, I had first looked at the effectors related to height and couldn't find anything wrong with them. FaceContractility is where I ended up stuck and confused, and it's also the only one where I can't follow the maths that implies that the derivative of that energy function is that gradient function.

In any case, thanks for getting back to me. Good to know that it does seem to work for some people :)

```
import matplotlib.pyplot as plt
%matplotlib inline
from tyssue import Sheet, SheetGeometry
from tyssue.generation import three_faces_sheet
from tyssue.draw.plt_draw import plot_forces
from tyssue.solvers.viscous import EulerSolver
from tyssue.dynamics import effectors
from tyssue.dynamics.factory import model_factory

smodel = model_factory(
    [effectors.FaceContractility],
    effectors.FaceContractility,
)
sheet = Sheet('3', *three_faces_sheet())
geom  = SheetGeometry
geom.update_all(sheet)

sheet.vert_df['viscosity'] = 2.0
sheet.face_df["contractility"] = 0.02

solver = EulerSolver(sheet, geom, smodel, auto_reconnect=True)

#fig, ax = plot_forces(sheet, geom, smodel, ['x', 'y'], 1)
res = solver.solve(35,1)
fig, ax = plot_forces(sheet, geom, smodel, ['x', 'y'], 1)

```

**Comment by glyg** [comment]
Hi 

Sorry for your troubles @btbest , thanks for your input @sniffo 

We use Euler solver for our latest models without (too much) issues, but it is prone to instability (due to the very crude ODE solving algo).

Good thing to do:
* scale the epithelium so the mean area of cell is 1:
`geom.scale(sheet, sheet.face_df["area"]**(-0.5), sheet.coords)`
* Then use Farhadifar et al values for line tension and contractility (0?12 and 0.04 respectively)
* Then you can increase contractility locally to have a cell contracting

I think (@sophietheis can confirm) that we use a small time step (like 0.1 or 0.01) to avoid divergences. To avoid too rapid movement without a lower timestep you can equivalently increase vertex viscosity

If you give us  some code maybe we can work this out?


Guillaume

**Comment by sophietheis** [comment]
Hi, 

In my model I use 1.0 at viscosity value and 0.1 as dt value. Also remove the autoreconnection help to understand what is not working. In your short example I try to take a bigger tissue and contract only the more central cell to see if it works and if it contracts more isotropically. 

Sophie.

**Comment by btbest** [comment]
Hi everyone,

thanks so much for your efforts and feedback! This was sort of a last-ditch attempt at figuring out where to go next, and you have helped me put together some advice for a potential successor to the project - since I'm no longer working on it. We have managed to get some extremely useful results with the quasistatic implementation; though the viscous paradigm would have been a lovely cherry on top since it is more intuitive for the biologists to think in time-resolved forces.

Your responses did manage to convince me that perhaps the effectors are just fine and the problem is elsewhere. I had previously been there, and tried to chance the `.solve()` call with different numbers or sizes of time steps. That hadn't helped, so I ended up back at figuring out which effector was erroneous. Your feedback sounded like you don't really modify the parameters during the simulation, which led me to try turning off the contractility-increasing mechanism built into the invagination model.

I ran it with a contractility value from close to the point where things spin out of control (400 - extremely high, this would be reached in sim cycle 32 normally, way after the invagination finishes in the original model) and it behaves perfectly fine; but even with such a ridiculous value, it took 150 time steps to constrict the mesoderm halfway. So I left the project with the conclusion that there are two problems to address for further work: 1) rechecking the unit scaling; there must be something really off for such high contractility values causing such slow movement and 2) rethinking if or how the effector parameters should be changed in the course of the sim to reflect the relevant biology.

So with that, thanks for building this amazing tool! I had a great 1.5 years on this project, and it was a lot of fun working with tyssue.

Best,
Bene
